src/docbuilder/limits.rs

```diff
@@ -0,0 +1,97 @@
+use error::Result;
+use postgres::Connection;
+use std::collections::BTreeMap;
+use std::time::Duration;
+
+const DEFAULT_MEMORY_LIMIT: usize = 3 * 1024 * 1024 * 1024; // 3 GB
+const DEFAULT_TIMEOUT: Duration = Duration::from_secs(15 * 60); // 15 minutes
+const DEFAULT_NETWORKING: bool = false;
+const DEFAULT_MAX_LOG_SIZE: usize = 100 * 1024; // 100 KB
+
+pub(crate) struct Limits {
+    memory: usize,
+    timeout: Duration,
+    networking: bool,
+    max_log_size: usize,
+}
+
+impl Default for Limits {
+    fn default() -> Self {
+        Self {
+            memory: DEFAULT_MEMORY_LIMIT,
+            timeout: DEFAULT_TIMEOUT,
+            networking: DEFAULT_NETWORKING,
+            max_log_size: DEFAULT_MAX_LOG_SIZE,
+        }
+    }
+}
+
+impl Limits {
+    pub(crate) fn for_crate(conn: &Connection, name: &str) -> Result<Self> {
+        let mut limits = Self::default();
+
+        let res = conn.query(
+            "SELECT * FROM sandbox_overrides WHERE crate_name = $1;",
+            &[&name],
+        )?;
+        if !res.is_empty() {
+            let row = res.get(0);
+            if let Some(memory) = row.get::<_, Option<i32>>("max_memory_bytes") {
+                limits.memory = memory as usize;
+            }
+            if let Some(timeout) = row.get::<_, Option<i32>>("timeout_seconds") {
+                limits.timeout = Duration::from_secs(timeout as u64);
+            }
+        }
+
+        Ok(limits)
+    }
+
+    pub(crate) fn memory(&self) -> usize {
+        self.memory
+    }
+
+    pub(crate) fn timeout(&self) -> Duration {
+        self.timeout
+    }
+
+    pub(crate) fn networking(&self) -> bool {
+        self.networking
+    }
+
+    pub(crate) fn max_log_size(&self) -> usize {
+        self.max_log_size
+    }
+
+    pub(crate) fn for_website(&self) -> BTreeMap<String, String> {
+        let time_scale = |v| scale(v, 60, &["seconds", "minutes", "hours"]);
+        let size_scale = |v| scale(v, 1024, &["bytes", "KB", "MB", "GB"]);
+
+        let mut res = BTreeMap::new();
+        res.insert("Available RAM".into(), size_scale(self.memory));
+        res.insert(
+            "Maximum rustdoc execution time".into(),
+            time_scale(self.timeout.as_secs() as usize),
+        );
+        res.insert("Maximum size of a build log".into(), size_scale(self.max_log_size));
+        if self.networking {
+            res.insert("Network access".into(), "allowed".into());
+        } else {
+            res.insert("Network access".into(), "blocked".into());
```

Comment:
All of the keys here are `&str` and that seems like it'll be true indefinitely, so maybe `BTreeMap<&'static str, String>` might make some sense?

Reply:
Heh, that's what I originally tried.

`rustc-serialize`'s [ToJson](https://docs.rs/rustc-serialize/0.3.24/rustc_serialize/json/trait.ToJson.html) is implemented for `BTreeMap<String, T>` but not `BTreeMap<&str, T>`, so this isn't really actionable without switching to serde (and changing this to serde would require a switch at least for all the frontend).